drivers/usbc/tcpc/ucpd_stm32.c

```diff
@@ -0,0 +1,1463 @@
+			}
+		} else if (info->evt & UCPD_EVT_RX_MSG) {
+			/*
+			 * In the case of a collsion, it's possible the port
+			 * partner may not send a GoodCRC and instead send the
+			 * message that was colliding. If a message is received
+			 * in this state, then treat it as a discard from an
+			 * incoming message.
+			 */
+			info->handler(info->dev, info->data,
+				      TCPC_ALERT_TRANSMIT_MSG_DISCARDED);
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+		}
+		break;
+
+	case STATE_HARD_RESET:
+		if (info->evt & UCPD_EVT_HR_DONE) {
+			/* HR complete, reset tx state values */
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+			data->ucpd_tx_request = 0;
+			data->tx_retry_count = 0;
+		} else if (info->evt & UCPD_EVT_HR_FAIL) {
+			ucpd_set_tx_state(info->dev, STATE_IDLE);
+			data->ucpd_tx_request = 0;
+			data->tx_retry_count = 0;
+		}
+		break;
+	}
+
+	/* If msg_src is valid, then start transmit */
+	if (msg_src > TX_MSG_NONE) {
+		ucpd_start_transmit(info->dev, msg_src);
+	}
+}
+
+/**
+ * @brief Alert handler
+ */
+static void ucpd_alert_handler(struct k_work *item)
+{
+	struct alert_info *info = CONTAINER_OF(item, struct alert_info, work);
+	struct tcpc_data *data = info->dev->data;
+
+	if (info->evt & UCPD_EVT_EVENT_CC) {
+		info->handler(info->dev, info->data, TCPC_ALERT_CC_STATUS);
+	}
+
+	if (info->evt & UCPD_EVT_HARD_RESET_RECEIVED) {
+		info->handler(info->dev, info->data,
+			      TCPC_ALERT_HARD_RESET_RECEIVED);
+	}
+
+	if (info->evt & UCPD_EVT_RX_MSG) {
+		info->handler(info->dev, info->data, TCPC_ALERT_MSG_STATUS);
+	}
+
+	/*
+	 * USB-PD messages are intiated in TCPM stack (PRL
+	 * layer). However, GoodCRC messages are initiated within the
+	 * UCPD driver based on USB-PD rx messages. These 2 types of
+	 * transmit paths are managed via events.
+	 *
+	 * UCPD generated GoodCRC messages, are the priority path as
+	 * they must be sent immediately following a successful USB-PD
+	 * rx message. As long as a transmit operation is not underway,
+	 * then a transmit message will be started upon request. The ISR
+	 * routine sets the event to indicate that the transmit
+	 * operation is complete.
+	 *
+	 * Hard reset requests are sent as a TCPM message, but in terms
+	 * of the ucpd transmitter, they are treated as a 3rd tx msg
+	 * source since they can interrupt an ongoing tx msg, and there
+	 * is no requirement to wait for a GoodCRC reply message.
+	 */
+
+	if (info->evt & UCPD_EVT_GOOD_CRC_REQ) {
+		data->ucpd_tx_request |= MSG_GOOD_CRC_MASK;
+	}
+
+	if (info->evt & UCPD_EVT_TCPM_MSG_REQ) {
+		data->ucpd_tx_request |= MSG_TCPM_MASK;
+	}
+
+	/*
+	 * Manage PD tx messages. The state machine may need to be
+	 * called more than once. For instance, if
+	 * the task is woken at the completion of sending a GoodCRC,
+	 * there may be a TCPM message request pending and just changing
+	 * the state back to idle would not trigger start of transmit.
+	 */
+	do {
+		ucpd_manage_tx(info);
+		/* Look at events only once. */
+		info->evt = 0;
```

Comment:
evt variable getting updated in interrupt context and worker context(thread)
if this variable get updated in isr while worker in this loop, that updated flags will get cleared here.
my recommendation is to clear only the flags which are processed above in an atomic way.

Reply:
I should have spotted this. Thanks

Reply:
I should have noticed this. Thanks